archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,718 @@
+/-
+Copyright (c) 2022. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching, Fabian Kruse, Nikolas Kuhn
+-/
+import analysis.p_series
+import analysis.special_functions.log.basic
+import analysis.special_functions.pow
+import algebra.big_operators.basic
+import algebra.big_operators.intervals
+import data.finset.sum
+import data.fintype.basic
+import data.real.basic
+import data.real.pi.wallis
+import order.filter
+import order.filter.basic
+import order.bounded_order
+import topology.instances.real
+import topology.instances.ennreal
+
+/-!
+# Stirling's formula
+
+This file proves Theorem 90 from the [100 Theorem List] <https://www.cs.ru.nl/~freek/100/>.
+It states that $n!$ grows asymptotically like $\sqrt{2\pi n}(\frac{n}{e})^n$.
+
+
+## Proof outline
+
+The proof follows: <https://proofwiki.org/wiki/Stirling%27s_Formula>.
+
+We proceed in two parts.
+### Part 1
+We consider the fraction sequence $a_n$ of fractions $n!$ over $\sqrt{2n}(\frac{n}{e})^n$ and
+proves that this sequence converges against a real, positve number $a$. For this the two main
+ingredients are
+ - taking the logarithm of the sequence and
+ - use the series expansion of $\log(1 + x)$.
+
+### Part 2
+We use the fact that the series defined in part 1 converges againt a real number $a$ and prove that
+$a = \sqrt{\pi}$. Here the main ingredient is the convergence of the Wallis product.
+-/
+
+
+open_locale big_operators -- notation ∑ for finite sums
+open_locale classical real topological_space nnreal ennreal filter big_operators
+open  finset filter nat real
+
+namespace stirling
+
+
+lemma partial_sum_le_tsum {a: ℕ → ℝ} (h : summable a) (g: ∀ (n:ℕ), 0 ≤ a n):
+∃ (c:ℝ), ∀ (s : finset ℕ), s.sum a ≤ c :=
```

Comment:
This lemma is a bit too specialized.  I eliminated the need for it in my golf below.